### DISCUSSION TITLE
A module for parsing versions (PEP-440)

### DISCUSSION THREAD

**Discussion by Totktonada** [tarantool/tarantool#7006] [Ideas]
Authors of external tarantool modules often need to determine a version of tarantool either to enable/disable some functionality or for the sake of conditional testing. It is [easy](https://stackoverflow.com/a/65330537/1598057) to obtain the version string, but splitting to components is implemented in modules again and again.

Examples of tarantool version parsing code:

* [tarantool-php (in tests)](https://github.com/tarantool/tarantool-php/blob/a631b56eba38ec3572c393b87ab54edf14b38aa2/test/shared/box.lua#L13-L65) (needs update for 2.10+)
* [tuple-keydef (in tests)](https://github.com/tarantool/tuple-keydef/blob/946466de4a6044e0bfba30eae851206ce0ae572c/test/tuple_keydef.test.lua#L3-L50) ([needs update](https://github.com/tarantool/tuple-keydef/issues/21) for 2.10+)
* [crud (in sources)](https://github.com/tarantool/crud/blob/249690a93a96d06c8401f0987e59bcc10f4bd00a/crud/common/utils.lua#L243-L355)
* [crud (in tests)](https://github.com/tarantool/crud/blob/249690a93a96d06c8401f0987e59bcc10f4bd00a/test/helper.lua#L303-L324)
* [cartridge-cli (in sources)](https://github.com/tarantool/cartridge-cli/commit/e37c98bac5db5fbb4fecb7a88233f95c2862de71)
* [vshard (in sources)](https://github.com/tarantool/vshard/commit/dbd2a2ebaf35120df287bbdefa518ed665f33a57)

I propose to extend the problem and implement a general version parsing code in tarantool. In particular, it should be applicable for parsing of tarantool's version.

Materials to dive into the topic:

* [Semantic Versioning](https://semver.org/)
* [PEP-440](https://peps.python.org/pep-0440/) ('Version Identification and Dependency Specification') and [pkg_resources.parse_version()](https://setuptools.pypa.io/en/latest/pkg_resources.html#parsing-utilities)
* [Debian Policy](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-standards-version) (see also [comments](https://github.com/knazarov/mkrepo/blob/8d96524bb65e87141fd7828255ed6860d06a71bf/debrepo.py#L439-L493) in mkrepo)
* [Fedora Versioning Guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/)
* Version [tokens](https://github.com/Homebrew/brew/blob/3acdab2d44f74f3cb838015bd41b92676965796d/Library/Homebrew/version.rb#L214-L332) and [parsers](https://github.com/Homebrew/brew/blob/3acdab2d44f74f3cb838015bd41b92676965796d/Library/Homebrew/version.rb#L383-L510) in `brew`
* [EAPI-7 functions](https://mgorny.pl/articles/the-ultimate-guide-to-eapi-7.html#version-manipulation-and-comparison-commands) for version manipulations in Gentoo

**Comment by ligurio** [comment]
BTW, there is a separate library [libversion](https://github.com/repology/libversion) that compares version in different formats.
